### PR TITLE
Feature/alexs changes

### PR DESCRIPTION
**Pull request by tavurth**
Also support better reward pipeline for later extension

Add more tests, and make tests more robust by implementing mocking.

Stop passing validator everywhere 🥳

### CODE REVIEW

File: neurons/miners/StableMiner/miner.py [hunk 1]
```diff
@@ -28,24 +29,47 @@ def __init__(self):
 
     def load_models(self):
         # Load the text-to-image model
-        self.t2i_model = AutoPipelineForText2Image.from_pretrained(
-            self.config.miner.model,
+        self.t2i_model_custom = AutoPipelineForText2Image.from_pretrained(
+            self.config.miner.custom_model,
```
Comment (Supreme-Emperor-Wang) [inline neurons/miners/StableMiner/miner.py]: Leave this as self.config.miner.model (the variable can be named model_custom though).

File: neurons/miners/StableMiner/miner.py [hunk 2]
```diff
@@ -28,24 +29,47 @@ def __init__(self):
...
         ).to(self.config.miner.device)
-        self.t2i_model.set_progress_bar_config(disable=True)
-        self.t2i_model.scheduler = DPMSolverMultistepScheduler.from_config(
-            self.t2i_model.scheduler.config
+        self.t2i_model_custom.set_progress_bar_config(disable=True)
+        self.t2i_model_custom.scheduler = DPMSolverMultistepScheduler.from_config(
+            self.t2i_model_custom.scheduler.config
+        )
+
+        self.t2i_model_alchemy = AutoPipelineForText2Image.from_pretrained(
```
Comment (Supreme-Emperor-Wang) [inline neurons/miners/StableMiner/miner.py]: Disable for now

File: neurons/miners/StableMiner/miner.py [hunk 3]
```diff
@@ -55,14 +79,28 @@ def load_models(self):
...
+                "args": self.i2i_args,
+                "model": self.i2i_model_custom,
+            },
         }
 
     def optimize_models(self):
         if self.config.miner.optimize:
-            self.t2i_model.unet = torch.compile(
-                self.t2i_model.unet, mode="reduce-overhead", fullgraph=True
+            self.t2i_model_alchemy.unet = torch.compile(
```
Comment (Supreme-Emperor-Wang) [inline neurons/miners/StableMiner/miner.py]: disable for now

File: neurons/utils.py [hunk 4]
```diff
@@ -446,7 +314,12 @@ def background_loop(self, is_validator):
 
                 if validator_weights:
                     weights_to_add = []
-                    for rw_name in self.reward_names:
+                    reward_names = [
+                        RewardModelType.IMAGE,
+                        RewardModelType.SIMILARITY,
```
Comment (Supreme-Emperor-Wang) [inline neurons/utils.py]: Comment out the Similarity model function

File: neurons/validator/forward.py [hunk 5]
```diff
@@ -1,63 +1,186 @@
...
+            batch_id=task.task_id,
+            prompt=task.prompt,
+            responses=[response],
+            masked_rewards=masked_rewards,
+        )
+        validator.batches_upload_queue.put_nowait(batch_for_upload)
+        responses.append(response)
+
+    # Responses with images should be first in list
+    responses.sort(key=lambda r: r.has_images(), reverse=True)
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/forward.py]: Double check that uids, hotkeys all match the change in order after the sort

File: neurons/validator/forward.py [hunk 6]
```diff
@@ -140,37 +263,93 @@ def gen_caption(prompt, i):
...
+            )
+            # im_bytes: image in binary format.
+            im_bytes = im_file.getvalue()
+            im_b64 = base64.b64encode(im_bytes)
+            images.append(im_b64.decode())
+            should_drop_entries.append(0)
+        else:
+            # Generated image has zero reward, we are dropping it
+            im_file = BytesIO()
+            T.transforms.ToPILImage()(
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/forward.py]: Are we just saving a blank image here? We may want to keep blacklisted images (but not NSFW ones).

File: neurons/validator/forward.py [hunk 7]
```diff
@@ -233,45 +398,58 @@ async def run_step(
...
-        validator, responses, uids, task_type
+    # Calculate rewards
+    automated_rewards: AutomatedRewards = await get_automated_rewards(
+        validator.model_type,
+        synapse,
+        responses,
+        task_type,
     )
 
-    scattered_rewards_adjusted = await get_human_rewards(validator, scattered_rewards)
```
Comment (Supreme-Emperor-Wang) [inline neurons/validator/forward.py:247]: Where do we retrieve the human rewards now?

File: neurons/miners/StableMiner/base.py [hunk 8]
```diff
@@ -397,14 +430,14 @@ def _base_priority(self, synapse) -> float:
         return priority
 
     def _base_blacklist(
-        self, synapse, vpermit_tao_limit=VPERMIT_TAO, rate_limit=1
+        self, synapse, vpermit_tao_limit=0.001, rate_limit=1
```
Comment (KMFODA) [inline neurons/miners/StableMiner/base.py]: Is this just for testing? If so we just need to remember to re-instate VPERMIT_TAO

Comment (tavurth) [reply neurons/miners/StableMiner/base.py]: Yep will do tomorrow

File: neurons/validator/forward.py (hunk 6 above)
Comment (tavurth) [reply neurons/validator/forward.py]: Hmm I'm not sure @1ex ?

File: neurons/validator/forward.py (hunk 7 above)
Comment (tavurth) [reply neurons/validator/forward.py:247]: in `human.py`, as part of standardised pipeline

**Pull request merged**